cpp/src/arrow/array/builder_binary.h

```diff
@@ -241,6 +252,13 @@ class ARROW_EXPORT FixedSizeBinaryBuilder : public ArrayBuilder {
   int32_t byte_width_;
   BufferBuilder byte_builder_;
 
+  template <typename Sized>
+  void CheckValueSize(const Sized& s, decltype(s.size())* = NULLPTR) {
```

Comment:
This is cryptic. Why didn't you keep the original solution?

Reply:
This allows CheckValueSize to be elided in release with only one `#ifndef NDEBUG` block, see this comment: https://github.com/apache/arrow/pull/3880#discussion_r266993207

The second function call is necessary because `DCHECK` can't go in headers

Reply:
Right, but my point is that this obscure. I'd rather keep the original solution.

Reply:
I'll revert it, then